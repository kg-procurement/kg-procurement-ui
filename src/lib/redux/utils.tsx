import { SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/query'

export type ZodParseError = Extract<
  FetchBaseQueryError,
  {
    status: 'PARSING_ERROR'
    originalStatus: number
    data: string
    error: string
  }
>

export function isZodParseError(
  err: SerializedError | FetchBaseQueryError | undefined,
): err is ZodParseError {
  if (!err) return false
  return (
    'data' in err &&
    'error' in err &&
    'originalStatus' in err &&
    'status' in err
  )
}

export function extractZodError(err: ZodParseError): React.ReactNode {
  const errorList = err.error.split('\n').slice(0, -1)
  return (
    <div className="flex flex-col">
      <div className="font-semibold">
        Response does not match validator schema:
      </div>
      <ul className="list-inside list-disc">
        {errorList.map((e, i) => (
          <li key={i}>{e}</li>
        ))}
      </ul>
    </div>
  )
}

export function isErrorWithMessage(
  err: FetchBaseQueryError,
): err is { status: number, data: { error: string } } {
  if (!err.data) return false
  return (
    typeof err.data === 'object' &&
    'error' in err.data &&
    typeof err.data.error === 'string'
  )
}
