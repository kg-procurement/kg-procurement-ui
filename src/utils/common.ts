export const noop = () => {}

export const formatPrice = (
  amount: number,
  quantityPrice: number,
  currencyCode: string,
  uomName: string
): string => {
  const formattedPrice = new Intl.NumberFormat(
    currencyCode === 'IDR' ? 'id-ID' : 'en-US',
    {
      style: 'currency',
      currency: currencyCode,
    }
  ).format(amount);

  return quantityPrice > 1
    ? `${formattedPrice} / ${quantityPrice} ${uomName}`
    : `${formattedPrice} / ${uomName}`;
};
