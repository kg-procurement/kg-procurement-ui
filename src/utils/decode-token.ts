export interface Claims {
  sub: string
  exp: number
}

export function decodeToken(token: string): Claims | null {
  try {
    const payload = JSON.parse(atob(token.split('.')[1]))
    return {
      sub: payload.sub,
      exp: payload.exp,
    }
  }
  catch (error) {
    console.error('Failed to decode token:', error)
    throw error
  }
}
