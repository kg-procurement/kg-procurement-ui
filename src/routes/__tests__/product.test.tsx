import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { withWrappers } from '@/lib/testing/utils.tsx'

import ProductPage from '../product.tsx'

describe('<ProductPage />', () => {
  it('should render the header section with the logo', () => {
    render(withWrappers(<ProductPage />))

    const logo = screen.getByAltText('Kompas Gramedia Logo Background')
    expect(logo).toBeInTheDocument()
    expect(logo).toHaveAttribute('src', '/kompas-gramedia-logo-bg.svg')

    expect(screen.getByText('Search Product')).toBeInTheDocument()
    expect(screen.getByText('Lorem Ipsum')).toBeInTheDocument()
  })

  it('should render inputs for filtering products', () => {
    render(withWrappers(<ProductPage />))

    screen.debug(undefined, 100000)
    const productNameInput = screen.getByPlaceholderText('Filter by Product')
    const locationDropdown = screen.getByRole('combobox')

    expect(productNameInput).toBeInTheDocument()
    expect(locationDropdown).toBeInTheDocument()
  })

  it('should correctly render table header content ', () => {
    render(withWrappers(<ProductPage />))

    const productTable = screen.getByRole('table')
    expect(productTable).toBeInTheDocument()

    expect(screen.getByText('Product Name')).toBeInTheDocument()
    expect(screen.getByText('Description')).toBeInTheDocument()
    expect(screen.getByText('Last Modified')).toBeInTheDocument()
  })

  it('should correctly render table header content ', () => {
    render(withWrappers(<ProductPage />))

    const productTable = screen.getByRole('table')
    expect(productTable).toBeInTheDocument()

    expect(screen.getByText('Product Name')).toBeInTheDocument()
    expect(screen.getByText('Description')).toBeInTheDocument()
    expect(screen.getByText('Last Modified')).toBeInTheDocument()
  })

  it('should correctly handle choose all product by checkbox on table header', async () => {
    render(withWrappers(<ProductPage />))

    const checkbox = screen.getAllByRole('checkbox')[0]
    await userEvent.click(checkbox)
    await userEvent.click(checkbox)
  })

  it('should correctly handle choosing a product by checkbox on table row', async () => {
    render(withWrappers(<ProductPage />))

    const checkbox = screen.getAllByRole('checkbox')[1]
    await userEvent.click(checkbox)
    await userEvent.click(checkbox)
  })

  it('should select an option on dropdown', async () => {
    render(withWrappers(<ProductPage />))

    const dropdown = screen.getByRole('combobox')
    await userEvent.click(dropdown)

    const option = screen.getByText('hello')
    await userEvent.click(option)
  })

  it('should render the footer', () => {
    render(withWrappers(<ProductPage />))
    expect(screen.getByText(/Â© 2024 KOMPAS/i)).toBeInTheDocument()
  })
})
