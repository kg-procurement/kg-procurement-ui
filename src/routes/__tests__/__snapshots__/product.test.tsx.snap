// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProductPage /> > should filter by product name is filled successfully 1`] = `"Product NameDescriptionPriceLast ModifiedBukuBukuRp 23.000,002020-11-11T13:22:16Zundefined"`;

exports[`<ProductPage /> > should handle when filter by product name returns null 1`] = `"Product NameDescriptionPriceLast Modified"`;

exports[`<ProductPage /> > should render the product vendors table content properly 1`] = `"Product NameDescriptionPriceLast ModifiedBukuBukuRp 23.000,002020-11-11T13:22:16ZundefinedKoranKoranRp 290.000,002020-11-02T14:49:06ZundefinedMajalahMajalahRp 450.000,002020-11-04T11:23:28Zundefined"`;
